### PR TITLE
Feature SDL-0273 WebEngine Projection mode

### PR DESCRIPTION
**Pull request by IGapchuk**
Implements [FORDTCN-4165](https://adc.luxoft.com/jira/browse/FORDTCN-4165)

This PR is **is ready** for internal review.

### Risk
This PR makes **minor** API changes.

### Summary
This PR provides the changes that allow using the WebEngine app's web view for the HMI in addition to the other available SDL templates.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/include/application_manager/commands/command_notification_impl.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ class CommandNotificationImpl : public CommandImpl {
   virtual bool Init();
   virtual bool CleanUp();
   virtual void Run();
-  void SendNotification();
+  void SendNotification(const bool final_message = false);
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/include/application_manager/commands/command_notification_impl.h:54]: New parameter require description

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_app_interface_unregistered_notification.cc [hunk 2]
```diff
@@ -57,7 +57,17 @@ OnAppInterfaceUnregisteredNotification::
 void OnAppInterfaceUnregisteredNotification::Run() {
   LOG4CXX_AUTO_TRACE(logger_);
 
-  SendNotification();
+  const bool final_message = IsFinalMessage();
+  SendNotification(final_message);
+}
+
+bool OnAppInterfaceUnregisteredNotification::IsFinalMessage() const {
+  using Reason = mobile_apis::AppInterfaceUnregisteredReason::eType;
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_app_interface_unregistered_notification.cc:65]: Interesting approach creating type shorter, i like it 
` using Reason = mobile_apis::AppInterfaceUnregisteredReason::eType;`

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 3]
```diff
@@ -1192,7 +1209,23 @@ mobile_apis::Result::eType RegisterAppInterfaceRequest::CheckWithPolicyData() {
         message[strings::msg_params][strings::app_hmi_type];
 
     AppHMITypeInserter inserter(app_hmi_type);
-    std::for_each(app_hmi_types.begin(), app_hmi_types.end(), inserter);
+    std::for_each(app_hmi_types_in_policy.begin(),
+                  app_hmi_types_in_policy.end(),
+                  inserter);
+  } else {
+    if (message[strings::msg_params].keyExists(strings::app_hmi_type)) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1216]: I would extract this if\else to functions, hard to read

File: src/components/application_manager/src/state_controller_impl.cc [hunk 4]
```diff
@@ -88,7 +88,10 @@ void StateControllerImpl::SetRegularState(ApplicationSharedPtr app,
     return;
   }
 
-  if (app->is_resuming() && !IsResumptionAllowed(app, state)) {
+  const bool app_is_resuming = app->is_resuming();
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/state_controller_impl.cc:91]: Extract it to separate commit if possible

File: src/components/application_manager/src/state_controller_impl.cc [hunk 5]
```diff
@@ -1268,7 +1275,8 @@ mobile_apis::VideoStreamingState::eType StateControllerImpl::CalcVideoState(
     ApplicationSharedPtr app,
     const mobile_apis::HMILevel::eType hmi_level) const {
   auto state = mobile_apis::VideoStreamingState::NOT_STREAMABLE;
-  if (IsStreamableHMILevel(hmi_level) && app->IsVideoApplication()) {
+
+  if (app->IsVideoApplication() && IsStreamableHMILevel(hmi_level)) {
```
Comment (LuxoftAKutsan) [inline src/components/application_manager/src/state_controller_impl.cc:1279]: What is the reason of order change?

File: src/components/protocol_handler/src/protocol_handler_impl.cc [hunk 6]
```diff
@@ -1106,6 +1108,24 @@ void ProtocolHandlerImpl::OnTMMessageSendFailed(
                                    << "bytes failed, connection_key "
                                    << message->connection_key()
                                    << "Error_text: " << error.text());
+
+  uint32_t connection_handle = 0;
```
Comment (LuxoftAKutsan) [inline src/components/protocol_handler/src/protocol_handler_impl.cc:1112]: This changes also need to be in separate commit

File: src/components/application_manager/test/state_controller/state_controller_test.cc [hunk 7]
```diff
@@ -3388,6 +3464,29 @@ TEST_F(
...
+  message[am::strings::msg_params][am::hmi_notification::is_active] = true;
+  message[am::strings::msg_params][am::hmi_notification::event_name] =
+      hmi_apis::Common_EventTypes::EMBEDDED_NAVI;
+  embedded_navi_event.set_smart_object(message);
+  state_ctrl_->on_event(embedded_navi_event);
+
+  EXPECT_CALL(*media_wep_app_ptr_, is_resuming())
+      .Times(2)
+      .WillRepeatedly(Return(true));
+  EXPECT_CALL(*media_wep_app_ptr_, SetRegularState(_, _)).Times(0);
```
Comment (IGapchuk) [inline src/components/application_manager/test/state_controller/state_controller_test.cc:3482]: To wich method's call these expectations are related: to `FullAudibleState()` or to `state_ctrl_->SetRegularState`? If they related to the `state_ctrl_->SetRegularState` we need to move the 3484 and 3485 lines above of expectations (because in this case these lines don't affect the expectations and could be moved).

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 8]
```diff
@@ -1166,17 +1173,27 @@ mobile_apis::Result::eType RegisterAppInterfaceRequest::CheckWithPolicyData() {
...
-      CheckMissedTypes checker(app_hmi_types, log);
-      std::for_each(app_types.begin(), app_types.end(), checker);
+      CheckMissedTypes checker(app_hmi_types_in_policy, log);
+      std::for_each(
+          app_types_in_message.begin(), app_types_in_message.end(), checker);
       if (!log.empty()) {
+        auto it = std::find(app_types_in_message.begin(),
+                            app_types_in_message.end(),
+                            mobile_apis::AppHMIType::WEB_VIEW);
+        if (app_types_in_message.end() != it) {
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1191]: @IGapchuk 
If AppHMIType is not included in policy - allow any type, but this wrong for mobile_apis::AppHMIType::WEB_VIEW.
So we try finding mobile_apis::AppHMIType::WEB_VIEW and if is present(app_types_in_message.end() != it). We return mobile_apis::Result::DISALLOWED

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc [hunk 9]
```diff
@@ -789,6 +829,257 @@ TEST_F(RegisterAppInterfaceRequestTest,
...
+      (*response_to_mobile)[am::strings::msg_params][am::strings::result_code]
+          .asUInt());
+  EXPECT_EQ(mobile_apis::Result::WARNINGS, result_code);
+
+  const std::string expected_info_to_mobile =
+      "Following AppHmiTypes are not present in policy "
+      "table:PROJECTION,INFORMATION,COMMUNICATION,MESSAGING,";
+  EXPECT_EQ(expected_info_to_mobile,
+            (*response_to_mobile)[am::strings::msg_params][am::strings::info]
+                .asString());
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc:973]: @IGapchuk 
In this case, we test the successful registration of the application for several types of AppHMIType that are absent in the policy, when mobile_apis :: AppHMIType :: WEB_VIEW are not installed for this application. This part of testing logic which
commented above:https://github.com/LuxoftSDL/sdl_core/pull/49#discussion_r442184023

[the remark this answers is not in the document]
Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc:973]: @AByzhynar please advice

File: src/components/application_manager/src/state_controller_impl.cc (hunk 5 above)
Comment (AByzhynar) [reply src/components/application_manager/src/state_controller_impl.cc:1279]: @LuxoftAKutsan Just convenience and readability. If it is not Video app -> there is now sense to call the method which will determine if HMI level is streamable or not.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1191]: @SNiukalov YOu are right. Your logic is correct.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 10]
```diff
@@ -1192,7 +1209,23 @@ mobile_apis::Result::eType RegisterAppInterfaceRequest::CheckWithPolicyData() {
         message[strings::msg_params][strings::app_hmi_type];
 
     AppHMITypeInserter inserter(app_hmi_type);
-    std::for_each(app_hmi_types.begin(), app_hmi_types.end(), inserter);
+    std::for_each(app_hmi_types_in_policy.begin(),
+                  app_hmi_types_in_policy.end(),
+                  inserter);
+  } else {
+    if (message[strings::msg_params].keyExists(strings::app_hmi_type)) {
+      smart_objects::SmartArray app_types_in_message =
```
Comment (AByzhynar) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1217]: @IGapchuk Is it possible to use `auto` here instead of explicit type?

File: src/components/application_manager/test/state_controller/state_controller_test.cc (hunk 7 above)
Comment (AByzhynar) [reply src/components/application_manager/test/state_controller/state_controller_test.cc:3482]: @IGapchuk All variables must be defined as close as possible to their first usage. They are defined here intentionally exactly before their usage.

File: src/components/application_manager/src/state_controller_impl.cc (hunk 4 above)
Comment (IGapchuk) [reply src/components/application_manager/src/state_controller_impl.cc:91]: @LuxoftAKutsan It is already in the separate commit - 8b8216822ccc953dadf09ca51aae928e9f8edd01

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (AByzhynar) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc:973]: @LitvinenkoIra I consider thqat this test case is not related to feature at all. It just check basic AppHMI types processing. 
Bus if we decided to add it - we need to check the info string. 
@IGapchuk is wrong. We do have a requirement for info string [here](https://adc.luxoft.com/jira/browse/APPLINK-16307) but the representation of a string is a bit different

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                                                   mock_hmi_capabilities_,
+                                                   mock_policy_handler_));
+  }
+
+  am::commands::MessageSharedPtr message_;
+  std::unique_ptr<OnAppInterfaceUnregisteredNotification> command_;
+  MockAppPtr mock_app_;
+};
+
+const auto data_to_check = [](const smart_objects::SmartObject& so) {
```
Comment (YarikMamykin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc:81]: Not obvious name - instead of check this lambda contains some kind of conversion from `SmartObject` to `std::pair`.
Another point - lambda function is still a function, not a simple variable, so its naming should be appropriate.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+#include "application_manager/commands/commands_test.h"
+#include "application_manager/smart_object_keys.h"
+#include "interfaces/MOBILE_API.h"
+#include "smart_objects/smart_object.h"
+
+namespace test {
+namespace components {
+namespace commands_test {
+namespace mobile_commands_test {
+namespace on_app_interface_unregistered_notification {
```
Comment (YarikMamykin) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc:48]: Maybe it is better to rename this namespace to `on_app_interface_unregistered_notification_test` so it won't interfere original one?

File: src/components/application_manager/src/state_controller_impl.cc [hunk 13]
```diff
@@ -895,7 +902,7 @@ void StateControllerImpl::OnApplicationRegistered(
     const mobile_apis::HMILevel::eType default_level) {
   LOG4CXX_AUTO_TRACE(logger_);
 
-  // After app registration HMI level should be set for DEFAUL_WINDOW only
+  // After app registration HMI level should be set for DEFAULT_WINDOW only
```
Comment (YarikMamykin) [inline src/components/application_manager/src/state_controller_impl.cc:905]: :+1:

File: src/components/application_manager/include/application_manager/commands/command_notification_impl.h (hunk 1 above)
Comment (IGapchuk) [reply src/components/application_manager/include/application_manager/commands/command_notification_impl.h:61]: @LuxoftAKutsan Done in c0352a9a390d3c4cfe97d88a53af79ddb2eb9a33

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc (hunk 3 above)
Comment (IGapchuk) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @LuxoftAKutsan Done in 6ce4904dba86238ab19eb600440ee82f0dd81d37

File: src/components/application_manager/src/application_impl.cc [hunk 14]
```diff
@@ -221,6 +223,11 @@ void ApplicationImpl::ChangeSupportingAppHMIType() {
             array_app_types[i].asUInt())) {
       mobile_projection_enabled_ = true;
     }
+    if (mobile_apis::AppHMIType::WEB_VIEW ==
+        static_cast<mobile_apis::AppHMIType::eType>(
+            array_app_types[i].asUInt())) {
+      webengine_projection_enabled_ = true;
+    }
```
[the remark this answers is not in the document]
Comment (IGapchuk) [reply src/components/application_manager/src/application_impl.cc:229]: Done in 6ce4904dba86238ab19eb600440ee82f0dd81d37

File: src/components/protocol_handler/src/protocol_handler_impl.cc (hunk 6 above)
Comment (IGapchuk) [reply src/components/protocol_handler/src/protocol_handler_impl.cc:1112]: @LuxoftAKutsan  Done in 4b8fbb11f16569ce0079f368ae0aa95540ffc89f

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (IGapchuk) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/register_app_interface_request_test.cc]: @AByzhynar @SNiukalov @LitvinenkoIra Was made the solution to check if response info string contains the expected AppHMITypes (in string representation). Please, take a look at the 9558c5c1d1fc3884e8b2f3268d1f935648853b99

File: src/components/application_manager/test/application_manager_impl_test.cc [hunk 15]
```diff
@@ -327,6 +327,43 @@ class ApplicationManagerImplTest
   void CreatePendingApplication();
 #endif
 
+  void ExpectedHmiState(const ApplicationSharedPtr wep_app,
+                        const WindowID main_window_id,
+                        const mobile_apis::HMILevel::eType hmi_level,
+                        const mobile_apis::VideoStreamingState::eType vss,
+                        const mobile_apis::AudioStreamingState::eType ass,
```
[the remark this answers is not in the document]
Comment (IGapchuk) [reply src/components/application_manager/test/application_manager_impl_test.cc]: Done in 988ab2f3d1afdd25e0a70f711222247a24372613

File: src/components/application_manager/test/state_controller/state_controller_test.cc [hunk 16]
```diff
@@ -1601,10 +1660,15 @@ TEST_F(StateControllerImplTest,
   am::ApplicationSharedPtr app_moved_to_full;
   NiceMock<application_manager_test::MockApplication>* app_moved_to_full_mock;
 
-  app_in_full =
-      ConfigureApp(&app_in_full_mock, 1761, 15685, NOT_MEDIA, NOT_NAVI, NOT_VC);
-  app_moved_to_full = ConfigureApp(
-      &app_moved_to_full_mock, 1796, 30093, NOT_MEDIA, NOT_NAVI, NOT_VC);
+  app_in_full = ConfigureApp(
+      &app_in_full_mock, 1761, 15685, NOT_MEDIA, NOT_NAVI, NOT_VC, NOT_WEP);
```
[the remark this answers is not in the document]
Comment (IGapchuk) [reply src/components/application_manager/test/state_controller/state_controller_test.cc]: Done in de709914fcd53ac4226dfbdaf2fd26927d3072a4

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc (hunk 10 above)
Comment (IGapchuk) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @AByzhynar This `if/else` was remade in the https://github.com/LuxoftSDL/sdl_core/commit/6ce4904dba86238ab19eb600440ee82f0dd81d37

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc (hunk 11 above)
Comment (IGapchuk) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc]: @YarikMamykin Done in cfb245dd9b7c5922c4b0ab05ef3f2f25f25fd88b

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc (hunk 12 above)
Comment (IGapchuk) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/test/commands/mobile/on_app_interface_unregistered_notification_test.cc]: @YarikMamykin Also done in the cfb245dd9b7c5922c4b0ab05ef3f2f25f25fd88b

**Comment by LitvinenkoIra** [comment]
@AByzhynar feel free to make some minor changes, because @IGapchuk is working on another task now

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 17]
```diff
@@ -1163,36 +1170,17 @@ mobile_apis::Result::eType RegisterAppInterfaceRequest::CheckWithPolicyData() {
...
-        message[strings::msg_params][strings::app_hmi_type];
-
-    AppHMITypeInserter inserter(app_hmi_type);
-    std::for_each(app_hmi_types.begin(), app_hmi_types.end(), inserter);
+  if (!app_hmi_types_in_policy.empty()) {
+    result = ProcessingAppHMITypesPolicies(message, app_hmi_types_in_policy);
+  } else {
+    result = ProcessingAppHMITypesInMessage(
+        message,
+        std::string("WEB_VIEW AppHmiType is absent in application policies, "
```
Comment (LitvinenkoIra) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1181]: @SNiukalov @IGapchuk I thought we do this check for all apps, not only for WEB_VIEW. Not clear why do we need this info string here

Comment (AByzhynar) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1181]: @LitvinenkoIra Because WEB VIEW absense is critical-> app won't be registered.  WebView type has a logic which differs from the other App HMI types.

Comment (LitvinenkoIra) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc:1181]: @AByzhynar I understand that we need this info in the message, but I do not think that we need to pass it as a parameter here

Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @AByzhynar @LitvinenkoIra 
Fixed in:https://github.com/LuxoftSDL/sdl_core/pull/49/commits/594f1c62e074e20ac698d1cf7cc7b35b200d9e0d

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 18]
```diff
@@ -1530,6 +1518,83 @@ bool RegisterAppInterfaceRequest::GetDataOnSessionKey(
...
+    LOG4CXX_DEBUG(logger_, response_info_);
+    return mobile_apis::Result::DISALLOWED;
+  }
+
+  return result;
+}
+
+mobile_apis::Result::eType
+RegisterAppInterfaceRequest::ProcessingAppHMITypesPolicies(
+    ns_smart_device_link::ns_smart_objects::SmartObject& message,
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @AByzhynar 
Fixed in:https://github.com/LuxoftSDL/sdl_core/pull/49/commits/594f1c62e074e20ac698d1cf7cc7b35b200d9e0d

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 19]
```diff
@@ -1530,6 +1518,83 @@ bool RegisterAppInterfaceRequest::GetDataOnSessionKey(
   return true;
 }
 
+mobile_apis::Result::eType
+RegisterAppInterfaceRequest::ProcessingAppHMITypesInMessage(
+    smart_objects::SmartObject& message,
+    const std::string& info,
+    const std::string& log) {
+  const bool appHMITypes_exist =
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @AByzhynar 
Fixed in:https://github.com/LuxoftSDL/sdl_core/pull/49/commits/594f1c62e074e20ac698d1cf7cc7b35b200d9e0d

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc [hunk 20]
```diff
@@ -1530,6 +1518,83 @@ bool RegisterAppInterfaceRequest::GetDataOnSessionKey(
   return true;
 }
 
+mobile_apis::Result::eType
+RegisterAppInterfaceRequest::ProcessingAppHMITypesInMessage(
+    smart_objects::SmartObject& message,
```
[the remark this answers is not in the document]
Comment (SNiukalov) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/register_app_interface_request.cc]: @AByzhynar 
Fixed in:https://github.com/LuxoftSDL/sdl_core/pull/49/commits/594f1c62e074e20ac698d1cf7cc7b35b200d9e0d

**Comment by LitvinenkoIra** [comment]
internal review completed
